core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java

```diff
@@ -162,119 +148,138 @@ public int getTopologyId() {
-                        }
-                     });
-            } catch (CacheException e) {
-               log.failedLoadingKeysFromCacheStore(e);
-            }
+         AtomicReference<List<InternalCacheEntry<Object, Object>>> batchRef =
+            new AtomicReference<>(Collections.emptyList());
+         subscription =
+            Flowable.concat(publishDataContainerEntries(), publishStoreEntries())
+                    .buffer(chunkSize)
+                    .concatMapCompletable(
+                       // Send the previous batch, not the current one
+                       // This allows us to mark all the segments as finished in the same RPC with the last batch
+                       entries -> {
+                          List<InternalCacheEntry<Object, Object>> previousBatch = batchRef.getAndSet(entries);
+                          if (previousBatch.isEmpty())
+                             return Completable.complete();
+
+                          return completionStageToCompletable().apply(sendEntries(previousBatch, false));
+                       })
+                    .subscribe(() -> {
+                       // Send the remaining entries and mark all the segments as finished
+                       List<InternalCacheEntry<Object, Object>> previousBatch = batchRef.get();
+                       sendEntries(previousBatch, true)
+                          .whenComplete((ignored, throwable) -> {
+                             taskFuture.complete(null);
+                          });
+                    }, throwable -> {
+                       taskFuture.complete(null);
+                    });
+         if (cancelled) {
+            // In case cancel() was called before we set the subscription field
+            subscription.dispose();
          }
-
-         // send the last chunk of all segments
-         sendEntries(true);
       } catch (Throwable t) {
-         // ignore eventual exceptions caused by cancellation (have InterruptedException as the root cause)
-         if (isCancelled()) {
-            if (trace) {
-               log.tracef("Ignoring error in already cancelled transfer to node %s, segments %s", destination,
-                          segments);
-            }
-         } else {
-            log.failedOutBoundTransferExecution(t);
-         }
-      }
-      if (trace) {
-         log.tracef("Completed outbound transfer to node %s, segments %s", destination, segments);
+         taskFuture.complete(null);
       }
+      return taskFuture;
    }
 
-   private void sendEntry(InternalCacheEntry ice, int segmentId) {
-      // send if we have a full chunk
-      if (accumulatedEntries >= chunkSize) {
-         sendEntries(false);
-         accumulatedEntries = 0;
-      }
+   private Flowable<InternalCacheEntry<Object, Object>> publishDataContainerEntries() {
+      return Flowable.fromIterable(() -> dataContainer.iterator(segments))
+                     .flatMapMaybe(ice -> {
```

Comment:
Are you sure the Maybe allocations are not worth removing? That would require N Maybe allocations as well as invoking a `flatMap` method, which has its own complications and additional allocations.

Reply:
I was thinking 1 allocation per entry isn't that bad, especially since it's collected almost immediately. Reading `FlowableFlatMapMaybe` though, looks like you're right and it creates plenty of other objects.

Anyway, I went and changed things so that `flatMapMaybe()` isn't needed anyway :)